From 27a1c0e05501028510ab44783720a5127dd7a96e Mon Sep 17 00:00:00 2001
From: Robert Lai <c_rlai@codeaurora.org>
Date: Thu, 10 Sep 2015 17:14:43 -0700
Subject: [PATCH] sched: optimize power cost/freq estimation

At present, calculation of frequency estimation and power cost at the
projected frequency for a given task takes big portion of function
select_best_cpu() latency.  This is because function power_cost()
calculates and converts the task's demand to projected frequency to
estimated power cost from select_best_cpu() path.

Introduce pre-calculated power cost and freqeuency per demand table to use
pre-calculated power cost rather than calculating these on the fly to
optimize the function power_cost().  This change is solely for optimization
purpose.

CRs-fixed: 849655
Change-Id: I0972a0b99dad1115cd41fb85785e5afae7138fd1
Signed-off-by: Joonwoo Park <joonwoop@codeaurora.org>
---
 include/linux/sched/sysctl.h |   4 ++
 kernel/sched/core.c          | 109 +++++++++++++++++++++++++++++++++++++++++++
 kernel/sched/fair.c          |  38 ++++++++-------
 kernel/sched/sched.h         |  13 ++++++
 kernel/sysctl.c              |   2 +-
 5 files changed, 149 insertions(+), 17 deletions(-)

diff --git a/include/linux/sched/sysctl.h b/include/linux/sched/sysctl.h
index f0729ab..0ade5e4 100644
--- a/include/linux/sched/sysctl.h
+++ b/include/linux/sched/sysctl.h
@@ -111,6 +111,10 @@ extern int sched_boost_handler(struct ctl_table *table, int write,
 extern int sched_window_update_handler(struct ctl_table *table,
 		 int write, void __user *buffer, size_t *lenp, loff_t *ppos);
 
+extern int sched_enable_power_aware_handler(struct ctl_table *table,
+					    int write, void __user *buffer,
+					    size_t *lenp, loff_t *ppos);
+
 #ifdef CONFIG_SCHED_DEBUG
 static inline unsigned int get_sysctl_timer_migration(void)
 {
diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index acf3f3c..a0f31da 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -2027,6 +2027,86 @@ const char *sched_window_reset_reasons[] = {
 	"MIGRATION_FIXUP_CHANGE",
 	"FREQ_ACCOUNT_WAIT_TIME_CHANGE"};
 
+/* Should be called under rq lock held */
+static void update_power_cost_table(struct rq *rq)
+{
+	int i;
+	u64 max_demand;
+	unsigned int max_freq, freqs, load;
+	struct cpu_pwr_stats *per_cpu_info = get_cpu_pwr_stats();
+	struct cpu_pstate_pwr *costs;
+	struct hmp_power_cost_table *ptr;
+	struct hmp_power_cost *map;
+
+	ptr = &rq->pwr_cost_table;
+	if (!sysctl_sched_enable_power_aware ||
+	    (!per_cpu_info || !per_cpu_info[rq->cpu].ptable))
+		return;
+
+	freqs = per_cpu_info[rq->cpu].len;
+	BUG_ON(freqs <= 0);
+
+	if (!ptr->len) {
+		map = kmalloc(sizeof(*(ptr->map)) * freqs, GFP_ATOMIC);
+		BUG_ON(!map);
+	} else {
+		BUG_ON(freqs != ptr->len);
+		map = ptr->map;
+	}
+	max_demand = div64_u64(max_task_load(), max_possible_capacity);
+	max_demand *= rq->max_possible_capacity;
+
+	costs = per_cpu_info[rq->cpu].ptable;
+	max_freq = costs[freqs - 1].freq;
+	for (i = 0; i < freqs; i++) {
+		load = costs[i].freq * 100 / max_freq;
+		map[i].freq = costs[i].freq;
+		map[i].power_cost = &costs[i].power;
+		map[i].demand = div64_u64(max_demand * load, 100);
+	}
+
+	/*
+	 * power_cost() doesn't hold rq lock, so set the rq->pwr_cost_table
+	 * afterwards.
+	 */
+	ptr->map = map;
+	ptr->len = freqs;
+}
+
+int sched_enable_power_aware_handler(struct ctl_table *table, int write,
+				     void __user *buffer, size_t *lenp,
+				     loff_t *ppos)
+{
+	int ret;
+	int i;
+	unsigned long flags;
+	unsigned int *data = (unsigned int *)table->data;
+	unsigned int old_val = *data;
+
+	ret = proc_dointvec(table, write, buffer, lenp, ppos);
+	if (ret || !write)
+		goto done;
+
+	if (write && (old_val == *data))
+		goto done;
+
+	if (*data != 0 && *data != 1) {
+		ret = -EINVAL;
+		*data = old_val;
+		goto done;
+	}
+
+	if (*data == 1) {
+		for_each_possible_cpu(i) {
+			raw_spin_lock_irqsave(&cpu_rq(i)->lock, flags);
+			update_power_cost_table(cpu_rq(i));
+			raw_spin_unlock_irqrestore(&cpu_rq(i)->lock, flags);
+		}
+	}
+done:
+	return ret;
+}
+
 /* Called with IRQs enabled */
 void reset_all_window_stats(u64 window_start, unsigned int window_size)
 {
@@ -2071,6 +2151,9 @@ void reset_all_window_stats(u64 window_start, unsigned int window_size)
 		reset_cpu_hmp_stats(cpu, 1);
 
 		fixup_nr_big_small_task(cpu, 0);
+
+		if (window_size)
+			update_power_cost_table(rq);
 	}
 
 	if (sched_window_stats_policy != sysctl_sched_window_stats_policy) {
@@ -2466,6 +2549,8 @@ static int cpufreq_notifier_policy(struct notifier_block *nb,
 
 			if (mplsf > highest_mplsf)
 				highest_mplsf = mplsf;
+
+			update_power_cost_table(rq);
 		}
 	}
 
@@ -2507,6 +2592,19 @@ static int cpufreq_notifier_trans(struct notifier_block *nb,
 	return 0;
 }
 
+static int pwr_stats_ready_notifier(struct notifier_block *nb,
+				    unsigned long cpu, void *data)
+{
+	struct rq *rq = cpu_rq(cpu);
+	unsigned long flags;
+
+	raw_spin_lock_irqsave(&rq->lock, flags);
+	update_power_cost_table(rq);
+	raw_spin_unlock_irqrestore(&rq->lock, flags);
+
+	return 0;
+}
+
 static struct notifier_block notifier_policy_block = {
 	.notifier_call = cpufreq_notifier_policy
 };
@@ -2515,6 +2613,15 @@ static struct notifier_block notifier_trans_block = {
 	.notifier_call = cpufreq_notifier_trans
 };
 
+static struct notifier_block notifier_pwr_stats_ready = {
+	.notifier_call = pwr_stats_ready_notifier
+};
+
+int __weak register_cpu_pwr_stats_ready_notifier(struct notifier_block *nb)
+{
+	return 1;
+}
+
 static int register_sched_callback(void)
 {
 	int ret;
@@ -2529,6 +2636,8 @@ static int register_sched_callback(void)
 		ret = cpufreq_register_notifier(&notifier_trans_block,
 						CPUFREQ_TRANSITION_NOTIFIER);
 
+	register_cpu_pwr_stats_ready_notifier(&notifier_pwr_stats_ready);
+
 	return 0;
 }
 
diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index cf37c3e..dfce98e 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -1772,11 +1772,16 @@ unsigned int power_cost_at_freq(int cpu, unsigned int freq)
 		 * hungry. */
 		return cpu_rq(cpu)->max_possible_capacity;
 
-	if (!freq)
-		freq = min_max_freq;
-
 	costs = per_cpu_info[cpu].ptable;
 
+	if (!freq) {
+		freq = min_max_freq;
+	} else {
+		i = per_cpu_info[cpu].len / 2;
+		if (costs[i].freq > freq)
+			i = 0;
+	}
+
 	while (costs[i].freq != 0) {
 		if (costs[i].freq >= freq ||
 		    costs[i+1].freq == 0)
@@ -1791,24 +1796,25 @@ unsigned int power_cost_at_freq(int cpu, unsigned int freq)
  * the CPU. */
 static unsigned int power_cost(u64 task_load, int cpu)
 {
-	unsigned int task_freq, cur_freq;
+	int i, end;
 	struct rq *rq = cpu_rq(cpu);
-	u64 demand;
+	struct hmp_power_cost_table *ptr = &rq->pwr_cost_table;
 
-	if (!sysctl_sched_enable_power_aware)
+	if (!sysctl_sched_enable_power_aware || !ptr->len)
 		return rq->max_possible_capacity;
 
-	/* calculate % of max freq needed */
-	demand = task_load * 100;
-	demand = div64_u64(demand, max_task_load());
-
-	task_freq = demand * rq->max_possible_freq;
-	task_freq /= 100; /* khz needed */
+	/* do simple divide & conquer */
+	i = ptr->len / 2;
+	end = ptr->len;
+	if (!(ptr->map[i].demand < task_load))
+		i = 0;
 
-	cur_freq = rq->cur_freq;
-	task_freq = max(cur_freq, task_freq);
-
-	return power_cost_at_freq(cpu, task_freq);
+	for (; i < end; i++) {
+		if (task_load <= ptr->map[i].demand &&
+		    ptr->map[i].freq >= rq->cur_freq)
+			return *(ptr->map[i].power_cost);
+	}
+	return *(ptr->map[i - 1].power_cost);
 }
 
 static int best_small_task_cpu(struct task_struct *p, int sync)
diff --git a/kernel/sched/sched.h b/kernel/sched/sched.h
index 8fcd6db..0c6b2ae 100644
--- a/kernel/sched/sched.h
+++ b/kernel/sched/sched.h
@@ -251,6 +251,17 @@ struct hmp_sched_stats {
 	u64 cumulative_runnable_avg;
 };
 
+struct hmp_power_cost {
+	unsigned int freq;
+	unsigned int *power_cost;
+	u64 demand;
+};
+
+struct hmp_power_cost_table {
+	int len;
+	struct hmp_power_cost *map;
+};
+
 #endif
 
 /* CFS-related fields in a runqueue */
@@ -525,6 +536,8 @@ struct rq {
 	u64 avg_irqload;
 	u64 irqload_ts;
 
+	struct hmp_power_cost_table pwr_cost_table;
+
 #ifdef CONFIG_SCHED_FREQ_INPUT
 	unsigned int old_busy_time;
 #endif
diff --git a/kernel/sysctl.c b/kernel/sysctl.c
index 8f19045..7c3be9d 100644
--- a/kernel/sysctl.c
+++ b/kernel/sysctl.c
@@ -447,7 +447,7 @@ static struct ctl_table kern_table[] = {
 		.data		= &sysctl_sched_enable_power_aware,
 		.maxlen		= sizeof(unsigned int),
 		.mode		= 0644,
-		.proc_handler	= proc_dointvec_minmax,
+		.proc_handler	= sched_enable_power_aware_handler,
 		.extra1		= &zero,
 		.extra2		= &one,
 	},
-- 
1.9.1

